tcptracer-bpf.c

```diff
@@ -0,0 +1,942 @@
+		};
+		int i;
+		for (i = 0; i < TASK_COMM_LEN; i++) {
+			evt4.comm[i] = p.comm[i];
+		}
+
+		u32 cpu;
+		cpu = bpf_get_smp_processor_id();
+		bpf_perf_event_output(ctx, &tcp_event_ipv4, cpu, &evt4, sizeof(evt4));
+		bpf_map_delete_elem(&tuplepid_ipv4, &t);
+	} else if (check_family(skp, AF_INET6)) {
+		u64 saddr_h, saddr_l, daddr_h, daddr_l;
+		saddr_h = 0;
+		saddr_l = 0;
+		daddr_h = 0;
+		daddr_l = 0;
+
+		bpf_probe_read(&daddr_h, sizeof(daddr_h), ((char *)skp) + status->offset_daddr_ipv6);
+		bpf_probe_read(&daddr_l, sizeof(daddr_l), ((char *)skp) + status->offset_daddr_ipv6 + sizeof(u64));
+		bpf_probe_read(&saddr_h, sizeof(saddr_h), ((char *)skp) + status->offset_daddr_ipv6 + 2 * sizeof(u64));
+		bpf_probe_read(&saddr_l, sizeof(saddr_l), ((char *)skp) + status->offset_daddr_ipv6 + 3 * sizeof(u64));
+		bpf_probe_read(&sport, sizeof(sport), ((char *)skp) + status->offset_sport);
+		bpf_probe_read(&dport, sizeof(dport), ((char *)skp) + status->offset_dport);
+
+		// if addresses or ports are 0, ignore
+		if ((saddr_h || saddr_l) == 0 || (daddr_h || daddr_l) == 0 || sport == 0 || dport == 0 ) {
+			return 0;
+		}
+
+		struct ipv6_tuple_t t = {
+			t.saddr_h = saddr_h,
+			t.saddr_l = saddr_l,
+			t.daddr_h = daddr_h,
+			t.daddr_l = daddr_l,
+			t.sport = ntohs(sport),
+			t.dport = ntohs(dport),
+			t.netns = net_ns_inum,
+		};
+
+		struct pid_comm *pp;
+		pp = bpf_map_lookup_elem(&tuplepid_ipv6, &t);
+		if (pp == 0) {
+			return 0;       // missed entry
+		}
+		struct pid_comm p = { };
+		bpf_probe_read(&p, sizeof(struct pid_comm), pp);
+		struct tcp_ipv6_event_t evt6 = {
+			.timestamp = bpf_ktime_get_ns(),
+			.cpu = bpf_get_smp_processor_id(),
+			.type = TCP_EVENT_TYPE_CONNECT,
+			.pid = p.pid >> 32,
+			.saddr_h = saddr_h,
+			.saddr_l = saddr_l,
+			.daddr_h = daddr_h,
+			.daddr_l = daddr_l,
+			.sport = ntohs(sport),
+			.dport = ntohs(dport),
+			.netns = net_ns_inum,
+		};
+		int i;
+		for (i = 0; i < TASK_COMM_LEN; i++) {
+			evt6.comm[i] = p.comm[i];
+		}
+
+		u32 cpu;
+		cpu = bpf_get_smp_processor_id();
+		bpf_perf_event_output(ctx, &tcp_event_ipv6, cpu, &evt6, sizeof(evt6));
+		bpf_map_delete_elem(&tuplepid_ipv6, &t);
+	}
+
+	return 0;
+}
+
+SEC("kprobe/tcp_close")
+int kprobe__tcp_close(struct pt_regs *ctx)
+{
+	struct sock *sk;
+	struct tcptracer_status_t *status;
+	u64 zero = 0;
+	u64 pid = bpf_get_current_pid_tgid();
+	u32 cpu = bpf_get_smp_processor_id();
+	sk = (struct sock *) PT_REGS_PARM1(ctx);
+
+	status = bpf_map_lookup_elem(&tcptracer_status, &zero);
+	if (status == NULL || status->status != TCPTRACER_STATUS_READY) {
+		return 0;
+	}
+
+	u32 net_ns_inum;
+	u16 sport, dport;
+	sport = 0;
+	dport = 0;
+
+	// Get network namespace id
+	possible_net_t *skc_net;
+
+	skc_net = NULL;
+	net_ns_inum = 0;
+	bpf_probe_read(&skc_net, sizeof(possible_net_t *), ((char *)sk) + status->offset_netns);
+	bpf_probe_read(&net_ns_inum, sizeof(net_ns_inum), ((char *)skc_net) + status->offset_ino);
+
+	if (check_family(sk, AF_INET)) {
```

Comment:
I did some refactoring, I didn't factor out the branches fully but it's more readable now, let me know if you want me to refactor more